### ISSUE TITLE
Keras on_epoch_end() callback output happens BEFORE completion of epoch output with verbose=1

### ISSUE THREAD

**Issue by poedator** [tensorflow/tensorflow#43184]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): modified TF example code
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Colab and Win10
- TensorFlow version (use command below): 2.3.0

**Describe the current behavior**
when training Model using `Model.fit()` with `verbose=1 `(for progress bar) and `keras.callbacks.Callback()` with `on_epoch_end() `function I observe that callback output happens BEFORE completion of the epoch progress output, like this:
```
Epoch 1/2
 63/125 [==============>...............] - ETA: 0s - loss: 43.2549 - mean_absolute_error: 4.6003
CALLBACK MESSAGE ON END EPOCH 0
125/125 [==============================] - 0s 794us/step - loss: 37.6690 - mean_absolute_error: 4.6575
```

**Describe the expected behavior**
callback output should happen AFTER completion of the epoch progress output:
```
Epoch 1/2
125/125 [==============================] - 0s 900us/step - loss: 36.9897 - mean_absolute_error: 4.8264
CALLBACK MESSAGE ON END EPOCH 0
```

**Standalone code to reproduce the issue**
[see gist here](https://colab.research.google.com/gist/poedator/05d7c2f1ac4cc5de65706896113b8e4d/custom_callback.ipynb)

**Comment by bhack** [comment]
The print is before the end cause you have the weight update step. 
Check this https://github.com/tensorflow/tensorflow/issues/29964#issuecomment-505498333 and follow-up comments.

I suppose that if there is something missing or that you want to better notify to developers you can open a small documentation PR on these lines https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/engine/training.py#L858:L862

**Comment by bhack** [comment]
So if you see it is after validation and before the training end https://github.com/tensorflow/tensorflow/blob/v2.3.0/tensorflow/python/keras/engine/training.py#L1135-L1145

**Comment by poedator** [comment]
@bnack, thank you for the replies, but I am not quite getting it. From the code of `Model.fit()` I see that the callback is only called at the VERY END of epoch. So why is callback output happening BEFORE the output of the last few batches? Could you please explain?
I also noticed that this problem is absent when batch training time is much greater. Could it be the case that batch output and callback output are done in parallel and the batch output gets delayed somehow?

**Comment by poedator** [comment]
I'd like to propose a simple commit to make the standard ProgbarLogger come first in the list of callbacks. It requires changing [this line](https://github.com/tensorflow/tensorflow/blob/v2.3.0/tensorflow/python/keras/callbacks.py#L260)  from `self.callbacks.append(self._progbar)` to `self.callbacks.insert(0, self._progbar)`
Which TF branch shall I commit to?

**Comment by bhack** [comment]
Yes PR are on master. I don't remeber the logic about why the are appended. You can try to collect feedbacks on the PR

**Comment by poedator** [comment]
just made [pull request](https://github.com/tensorflow/tensorflow/pull/43477)

**Comment by amahendrakar** [comment]
@poedator,
Looks like the issue has been fixed. 

On running the code with the latest TF-nightly, I got the expected output. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/0eef7549a3524760e55de49a56801224/43184.ipynb). Thanks!

**Comment by poedator** [comment]
Thanks, guys! This is my first PR merged into a major project.
